src/snowflake/cli/api/utils/templating_functions.py

```diff
@@ -0,0 +1,85 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from typing import Any, List, Optional
+
+from snowflake.cli.api.exceptions import InvalidTemplate
+from snowflake.cli.api.project.util import (
+    concat_identifiers,
+    identifier_to_str,
+    to_identifier,
+)
+
+
+class TemplatingFunctions:
+    """
+    This class contains all the functions available for templating.
+    Any callable not starting with '_' will automatically be available for users to use.
+    """
+
+    @staticmethod
+    def _verify_str_arguments(
+        func_name: str,
+        args: List[Any],
+        *,
+        min_count: Optional[int] = None,
+        max_count: Optional[int] = None,
+    ):
+        if min_count is not None and len(args) < min_count:
+            raise InvalidTemplate(
+                f"{func_name} requires at least {min_count} argument(s)"
+            )
+
+        if max_count is not None and len(args) > max_count:
+            raise InvalidTemplate(
+                f"{func_name} supports at most {max_count} argument(s)"
+            )
+
+        for arg in args:
+            if not isinstance(arg, str):
+                raise InvalidTemplate(f"{func_name} only accepts String values")
+
+    @staticmethod
+    def id_concat(*args):
+        TemplatingFunctions._verify_str_arguments("id_concat", args, min_count=1)
+        return concat_identifiers(args)
+
+    @staticmethod
+    def str_to_id(*args):
+        TemplatingFunctions._verify_str_arguments(
+            "str_to_id", args, min_count=1, max_count=1
+        )
+        return to_identifier(args[0])
+
+    @staticmethod
+    def id_to_str(*args):
+        TemplatingFunctions._verify_str_arguments(
+            "id_to_str", args, min_count=1, max_count=1
+        )
+        return identifier_to_str(args[0])
+
+
+def get_templating_functions():
+    """
+    Returns a dictionary with all the functions available for templating
+    """
+    templating_functions = {
+        func: getattr(TemplatingFunctions, func)
+        for func in dir(TemplatingFunctions)
+        if callable(getattr(TemplatingFunctions, func)) and not func.startswith("_")
+    }
```

Comment:
Publicly exposed functions should be exported explicitly, by name. Something like:
```
 {
    "foo": foo,
    "bar": bar,
}
```

would be a lot clearer (and would make aliases trivial too if we ever need them). Or take a look at the @public scheme I'm using in the new repo, but IMHO it's a bit overkill here.

Reply:
COPY OF ANOTHER COMMENT:

From one of your previous comments, I got the idea that we want to avoid defining these in 2 places.

In this method, you only need to add the new function to this class.
Alternatively, we can define the functions separately and manually in a map:
```
def id_to_str(arg):
  ....


- Alternatively, we can do the following:
templating_functions = {
  'id_to_str': id_to_str
}
```
But then there is a bit of boilerplate as we would have to define/reference the function in 2 places. Not a big deal, but let me know what you think.

Reply:
Subjective. Will get behind this, although I would have probably gone with your alternative approach if I had done this work myself.